File: c10/core/DispatchKey.cpp
#include <c10/core/DispatchKey.h>
#include <c10/core/DispatchKeySet.h>

#include <unordered_map>

namespace c10 {

const char* toString(BackendComponent t) {
  switch (t) {
    case BackendComponent::CPUBit:
      return "CPUBit";
    case BackendComponent::CUDABit:
      return "CUDABit";
    case BackendComponent::HIPBit:
      return "HIPBit";
    case BackendComponent::XLABit:
      return "XLABit";
    case BackendComponent::LazyBit:
      return "LazyBit";
    case BackendComponent::MetaBit:
      return "MetaBit";
    case BackendComponent::XPUBit:
      return "XPUBit";
    case BackendComponent::IPUBit:
      return "IPUBit";
    case BackendComponent::MPSBit:
      return "MPSBit";
    case BackendComponent::HPUBit:
      return "HPUBit";
    case BackendComponent::VEBit:
      return "VEBit";
    case BackendComponent::MTIABit:
      return "MTIA";
    case BackendComponent::PrivateUse1Bit:
      return "PrivateUse1Bit";
    case BackendComponent::PrivateUse2Bit:
      return "PrivateUse2Bit";
    case BackendComponent::PrivateUse3Bit:
      return "PrivateUse3Bit";
    case BackendComponent::InvalidBit:
      return "InvalidBit";
    default:
      return "UNKNOWN_BACKEND_BIT";
  }
}

BackendComponent toBackendComponent(DeviceType device_type) {
  switch (device_type) {
#define DO_CASE(device, _)                          \
  case DeviceType::device: {                        \
    return toBackendComponent(DispatchKey::device); \
  }
    C10_FORALL_BACKEND_DEVICE_TYPES(DO_CASE, unused)
#undef DO_CASE
    default:
      return BackendComponent::InvalidBit;
  }
}

const char* toString(DispatchKey t) {
  switch (t) {
    case DispatchKey::Undefined:
      return "Undefined";

    case DispatchKey::Dense:
      return "Dense";
    case DispatchKey::FPGA:
      return "FPGA";
    case DispatchKey::ORT:
      return "ORT";
    case DispatchKey::Vulkan:
      return "Vulkan";
    case DispatchKey::Metal:
      return "Metal";

    case DispatchKey::Lazy:
      return "Lazy";
    case DispatchKey::MPS:
      return "MPS";
    case DispatchKey::HPU:
      return "HPU";
    case DispatchKey::MTIA:
      return "MTIA";

    case DispatchKey::Quantized:
      return "Quantized";
    case DispatchKey::CustomRNGKeyId:
      return "CustomRNGKeyId";
    case DispatchKey::MkldnnCPU:
      return "MkldnnCPU";

    case DispatchKey::Sparse:
      return "Sparse";
    case DispatchKey::SparseCsrCPU:
      return "SparseCsrCPU";
    case DispatchKey::SparseCsrCUDA:
      return "SparseCsrCUDA";

    case DispatchKey::NestedTensor:
      return "NestedTensor";

    case DispatchKey::BackendSelect:
      return "BackendSelect";

    case DispatchKey::Python:
      return "Python";

    case DispatchKey::Fake:
      return "Fake";
    case DispatchKey::FuncTorchDynamicLayerBackMode:
      return "FuncTorchDynamicLayerBackMode";

    case DispatchKey::Functionalize:
      return "Functionalize";

    case DispatchKey::Named:
      return "Named";

    case DispatchKey::Conjugate:
      return "Conjugate";
    case DispatchKey::Negative:
      return "Negative";
    case DispatchKey::ZeroTensor:
      return "ZeroTensor";

    case DispatchKey::ADInplaceOrView:
      return "ADInplaceOrView";

    case DispatchKey::AutogradOther:
      return "AutogradOther";
    case DispatchKey::AutogradFunctionality:
      return "AutogradFunctionality";
    case DispatchKey::AutogradNestedTensor:
      return "AutogradNestedTensor";

    case DispatchKey::Tracer:
      return "Tracer";

    case DispatchKey::AutocastCPU:
      return "AutocastCPU";
    case DispatchKey::AutocastXPU:
      return "AutocastXPU";
    case DispatchKey::AutocastIPU:
      return "AutocastIPU";
    case DispatchKey::AutocastHPU:
      return "AutocastHPU";
    case DispatchKey::AutocastCUDA:
      return "AutocastCUDA";
    case DispatchKey::AutocastPrivateUse1:
      return "AutocastPrivateUse1";

    case DispatchKey::FuncTorchBatched:
      return "FuncTorchBatched";
    case DispatchKey::FuncTorchVmapMode:
      return "FuncTorchVmapMode";

    case DispatchKey::Batched:
      return "Batched";
    case DispatchKey::VmapMode:
      return "VmapMode";

    case DispatchKey::FuncTorchGradWrapper:
      return "FuncTorchGradWrapper";

    case DispatchKey::DeferredInit:
      return "DeferredInit";
    case DispatchKey::PythonTLSSnapshot:
      return "PythonTLSSnapshot";

    // Note [Out-of-tree vmap+grad prototype]
    // The following keys are used in the implementation of the out-of-tree
    // composable functions transforms (vmap+grad) prototype that lives at
    // https://github.com/zou3519/functorch
    // We plan on eventually upstreaming the prototype into core, at which
    // point it will have a different design that should use fewer keys.
    case DispatchKey::FuncTorchDynamicLayerFrontMode:
      return "FuncTorchDynamicLayerFrontMode";

    case DispatchKey::TESTING_ONLY_GenericWrapper:
      return "TESTING_ONLY_GenericWrapper";

    case DispatchKey::TESTING_ONLY_GenericMode:
      return "TESTING_ONLY_GenericMode";

    case DispatchKey::PreDispatch:
      return "PreDispatch";

    case DispatchKey::PythonDispatcher:
      return "PythonDispatcher";

      // Aliases

    case DispatchKey::Autograd:
      return "Autograd";
    case DispatchKey::CompositeImplicitAutograd:
      return "CompositeImplicitAutograd";
    case DispatchKey::CompositeImplicitAutogradNestedTensor:
      return "CompositeImplicitAutogradNestedTensor";
    case DispatchKey::CompositeExplicitAutograd:
      return "CompositeExplicitAutograd";
    case DispatchKey::CompositeExplicitAutogradNonFunctional:
      return "CompositeExplicitAutogradNonFunctional";
    case DispatchKey::FuncTorchBatchedDecomposition:
      return "FuncTorchBatchedDecomposition";

      // Per-backend dispatch keys

    default:
      auto bc = toBackendComponent(t);
      auto fk = toFunctionalityKey(t);

      switch (fk) {
#define ENTRY(backend, functionality)  \
  case BackendComponent::backend##Bit: \
    return #functionality #backend;

#define FORALL_BC(dkname, prefix)                  \
  case DispatchKey::dkname:                        \
    switch (bc) {                                  \
      C10_FORALL_BACKEND_COMPONENTS(ENTRY, prefix) \
      default:                                     \
        return #prefix "Undefined";                \
    }

        C10_FORALL_FUNCTIONALITY_KEYS(FORALL_BC)

        default:
          switch (bc) {
            C10_FORALL_BACKEND_COMPONENTS(ENTRY, Unknown)
            default:
              return "UnknownUnknown";
          }

#undef FORALL_BC
#undef ENTRY
      }
  }
}

std::ostream& operator<<(std::ostream& str, DispatchKey rhs) {
  return str << toString(rhs);
}
std::ostream& operator<<(std::ostream& str, BackendComponent rhs) {
  return str << toString(rhs);
}

DispatchKey getAutogradKeyFromBackend(BackendComponent k) {
  // We want this to return an autograd key. We're relying on the fact that
  // getAutogradRelatedKeySetFromBackend returns an autograd key +
  // ADInplaceOrView, and autograd has higher precedence. The core mapping from
  // backend -> autograd key lives in `getAutogradRelatedKeySetFromBackend`
  // instead of here for performance. `getAutogradRelatedKeySetFromBackend` is a
  // hotpath function, and we want to make sure that it doesn't have to
  // construct any DispatchKeySets at runtime.
  return getAutogradRelatedKeySetFromBackend(k).highestPriorityTypeId();
}

c10::DispatchKey parseDispatchKey(const std::string& k) {
  static std::unordered_map<std::string, c10::DispatchKey> key_map = {
      {"Undefined", c10::DispatchKey::Undefined},
      {"Dense", c10::DispatchKey::Dense},
      {"FPGA", c10::DispatchKey::FPGA},
      {"ORT", c10::DispatchKey::ORT},
      {"MPS", c10::DispatchKey::MPS},
      {"Vulkan", c10::DispatchKey::Vulkan},
      {"Metal", c10::DispatchKey::Metal},
      {"VE", c10::DispatchKey::VE},
      {"Meta", c10::DispatchKey::Meta},
      {"Quantized", c10::DispatchKey::Quantized},
      {"CustomRNGKeyId", c10::DispatchKey::CustomRNGKeyId},
      {"MkldnnCPU", c10::DispatchKey::MkldnnCPU},
      {"Sparse", c10::DispatchKey::Sparse},
      {"SparseCsrCPU", c10::DispatchKey::SparseCsrCPU},
      {"SparseCsrCUDA", c10::DispatchKey::SparseCsrCUDA},
      {"BackendSelect", c10::DispatchKey::BackendSelect},
      {"Python", c10::DispatchKey::Python},
      {"PythonTLSSnapshot", c10::DispatchKey::PythonTLSSnapshot},
      {"Fake", c10::DispatchKey::Fake},
      {"Named", c10::DispatchKey::Named},
      {"Conjugate", c10::DispatchKey::Conjugate},
      {"Negative", c10::DispatchKey::Negative},
      {"ZeroTensor", c10::DispatchKey::ZeroTensor},
      {"FuncTorchDynamicLayerBackMode",
       c10::DispatchKey::FuncTorchDynamicLayerBackMode},
      {"Functionalize", c10::DispatchKey::Functionalize},
      {"ADInplaceOrView", c10::DispatchKey::ADInplaceOrView},
      {"AutogradOther", c10::DispatchKey::AutogradOther},
      {"AutogradFunctionality", c10::DispatchKey::AutogradFunctionality},
      {"AutogradNestedTensor", c10::DispatchKey::AutogradNestedTensor},
      {"Tracer", c10::DispatchKey::Tracer},
      {"AutocastCPU", c10::DispatchKey::AutocastCPU},
      {"AutocastXPU", c10::DispatchKey::AutocastXPU},
      {"AutocastIPU", c10::DispatchKey::AutocastIPU},
      {"AutocastHPU", c10::DispatchKey::AutocastHPU},
      {"AutocastCUDA", c10::DispatchKey::AutocastCUDA},
      {"AutocastPrivateUse1", c10::DispatchKey::AutocastPrivateUse1},
      {"FuncTorchBatched", c10::DispatchKey::FuncTorchBatched},
      {"FuncTorchVmapMode", c10::DispatchKey::FuncTorchVmapMode},
      {"Batched", c10::DispatchKey::Batched},
      {"VmapMode", c10::DispatchKey::VmapMode},
      {"DeferredInit", c10::DispatchKey::DeferredInit},
      {"FuncTorchGradWrapper", c10::DispatchKey::FuncTorchGradWrapper},
      {"FuncTorchDynamicLayerFrontMode",
       c10::DispatchKey::FuncTorchDynamicLayerFrontMode},
      {"TESTING_ONLY_GenericWrapper",
       c10::DispatchKey::TESTING_ONLY_GenericWrapper},
      {"TESTING_ONLY_GenericMode", c10::DispatchKey::TESTING_ONLY_GenericMode},
      {"PythonDispatcher", c10::DispatchKey::PythonDispatcher},
      {"PreDispatch", c10::DispatchKey::PreDispatch},

      {"CPU", c10::DispatchKey::CPU},
      {"CUDA", c10::DispatchKey::CUDA},
      {"HIP", c10::DispatchKey::HIP},
      {"XLA", c10::DispatchKey::XLA},
      {"MPS", c10::DispatchKey::MPS},
      {"XPU", c10::DispatchKey::XPU},
      {"IPU", c10::DispatchKey::IPU},
      {"HPU", c10::DispatchKey::HPU},
      {"Lazy", c10::DispatchKey::Lazy},
      {"MTIA", c10::DispatchKey::MTIA},
      {"NestedTensor", c10::DispatchKey::NestedTensor},
      {"NestedTensorCPU", c10::DispatchKey::NestedTensorCPU},
      {"NestedTensorCUDA", c10::DispatchKey::NestedTensorCUDA},
      {"NestedTensorMeta", c10::DispatchKey::NestedTensorMeta},
      {"NestedTensorPrivateUse1", c10::DispatchKey::NestedTensorPrivateUse1},
      {"PrivateUse1", c10::DispatchKey::PrivateUse1},
      {"PrivateUse2", c10::DispatchKey::PrivateUse2},
      {"PrivateUse3", c10::DispatchKey::PrivateUse3},

      {"QuantizedCPU", c10::DispatchKey::QuantizedCPU},
      {"QuantizedCUDA", c10::DispatchKey::QuantizedCUDA},
      {"QuantizedXPU", c10::DispatchKey::QuantizedXPU},
      {"QuantizedPrivateUse1", c10::DispatchKey::QuantizedPrivateUse1},

      {"SparseCPU", c10::DispatchKey::SparseCPU},
      {"SparseCUDA", c10::DispatchKey::SparseCUDA},
      {"SparseHIP", c10::DispatchKey::SparseHIP},
      {"SparseXPU", c10::DispatchKey::SparseXPU},
      {"SparseVE", c10::DispatchKey::SparseVE},
      {"SparseMeta", c10::DispatchKey::SparseMeta},
      {"SparsePrivateUse1", c10::DispatchKey::SparsePrivateUse1},

      {"AutogradCPU", c10::DispatchKey::AutogradCPU},
      {"AutogradCUDA", c10::DispatchKey::AutogradCUDA},
      {"AutogradXLA", c10::DispatchKey::AutogradXLA},
      {"AutogradLazy", c10::DispatchKey::AutogradLazy},
      {"AutogradMeta", c10::DispatchKey::AutogradMeta},
      {"AutogradIPU", c10::DispatchKey::AutogradIPU},
      {"AutogradXPU", c10::DispatchKey::AutogradXPU},
      {"AutogradMPS", c10::DispatchKey::AutogradMPS},
      {"AutogradHPU", c10::DispatchKey::AutogradHPU},
      {"AutogradPrivateUse1", c10::DispatchKey::AutogradPrivateUse1},
      {"AutogradPrivateUse2", c10::DispatchKey::AutogradPrivateUse2},
      {"AutogradPrivateUse3", c10::DispatchKey::AutogradPrivateUse3},

      {"Autograd", c10::DispatchKey::Autograd},
      {"CompositeImplicitAutograd",
       c10::DispatchKey::CompositeImplicitAutograd},
      {"CompositeImplicitAutogradNestedTensor",
       c10::DispatchKey::CompositeImplicitAutogradNestedTensor},
      {"CompositeExplicitAutograd",
       c10::DispatchKey::CompositeExplicitAutograd},
      {"CompositeExplicitAutogradNonFunctional",
       c10::DispatchKey::CompositeExplicitAutogradNonFunctional},
      {"FuncTorchBatchedDecomposition",
       c10::DispatchKey::FuncTorchBatchedDecomposition},
  };
  auto it = key_map.find(k);
  TORCH_CHECK(it != key_map.end(), "could not parse dispatch key: ", k);
  return it->second;
}

} // namespace c10

File: torch/_mock_dispatcher/dispatch_key.py
from enum import Enum, auto

class DispatchKey(Enum):
    # "Dense" backends
    # These are "real" backends that can also have different sparse/autograd/quantized kernels
    CPUBit = 0
    CUDABit = auto()
    HIPBit = auto()
    XLABit = auto()
    LazyBit = auto()
    XPUBit = auto()
    NestedTensorBit = auto()
    MLCBit = auto()
    HPUBit = auto()
    PrivateUse1Bit = auto()
    PrivateUse2Bit = auto()
    PrivateUse3Bit = auto()

    # EndOfExtensibleBackendKeys tells us which keys should get "backend bits" in the runtime bitset.
    EndOfBackendKeys = PrivateUse3Bit

    # Undefined technically counts as a "functionality" bit - it's the lowest priority functionality.
    Undefined = auto()
    CatchAll = Undefined


    # Dense lives above the "Other" backends, mostly because the other backends should have higher priority than dense CPU.
    # e.g. torch.add(vulkan_tensor, cpu_tensor) should dispatch to the vulkan kernel
    Dense = auto()

    # "Other" backends
    # These are backends in core that currently don't need separate implementations for sparse, autograd or quantized kernels
    # Note: this means that none of these backends can customize autograd. The "AutogradOther" bit will be added to their bitset.
    FPGA = auto()
    ORT = auto()
    Vulkan = auto()
    Metal = auto()
    MKLDNN = auto()
    OpenGL = auto()
    OpenCL = auto()
    IDEEP = auto()
    CustomRNGKeyId = auto()
    MkldnnCPU = auto()
    Meta = auto()

    # "Functionality" keys start here (except for Dense, which is further up).
    Quantized = auto()
    Sparse = auto()

    # We could consider making "SparseCsr" a per-backend concept
    SparseCsrCPU = auto()
    SparseCsrCUDA = auto()

    BackendSelect = auto()
    Named = auto()

    # We need AutogradOther because we don't currently give every backend an actual bit.
    # Every key between "EndOfBackendKeys" and "Dense" corresponds to a backend that still needs to work with autograd.
    AutogradOther = auto()
    Autograd = auto()

    Tracer = auto()
    Autocast = auto()
    Batched = auto()
    VmapMode = auto()
    TESTING_ONLY_GenericWrapper = auto()
    TESTING_ONLY_GenericMode = auto()
    EndOfFunctionalityKeys = TESTING_ONLY_GenericMode

    # Per-backend functionality keys
    CPU = auto()
    CUDA = auto()
    HIP = auto()
    XLA = auto()
    Lazy = auto()
    XPU = auto()
    NestedTensor = auto()
    MLC = auto()
    HPU = auto()
    PrivateUse1 = auto()
    PrivateUse2 = auto()
    PrivateUse3 = auto()

    QuantizedCPU = auto()
    QuantizedCUDA = auto()
    QuantizedXPU = auto()

    SparseCPU = auto()
    SparseCUDA = auto()
    SparseHIP = auto()
    SparseXPU = auto()

    AutogradCPU = auto()
    AutogradCUDA = auto()
    AutogradXLA = auto()
    AutogradLazy = auto()
    AutogradNestedTensor = auto()
    AutogradMLC = auto()
    AutogradHPU = auto()
    AutogradXPU = auto()
    AutogradPrivateUse1 = auto()
    AutogradPrivateUse2 = auto()
    AutogradPrivateUse3 = auto()
    EndOfPerBackendKeys = AutogradPrivateUse3

    AutogradAlias = auto()
    CompositeImplicitAutograd = auto()
    CompositeExplicitAutograd = auto()
    EndOfAliasKeys = CompositeExplicitAutograd

    CPUTensorId = CPU
    CUDATensorId = CUDA
    PrivateUse1_PreAutograd = AutogradPrivateUse1
    PrivateUse2_PreAutograd = AutogradPrivateUse2
    PrivateUse3_PreAutograd = AutogradPrivateUse3

    def __str__(self) -> str:
        return self.name

    def lower(self) -> str:
        return str(self).lower()

def isPerBackendFunctionalityKey(k: DispatchKey) -> bool:
    if k is DispatchKey.Dense or k is DispatchKey.Quantized or k is DispatchKey.Sparse or k is DispatchKey.Autograd:
        return True
    return False

def isRuntimeDispatchKey(k: DispatchKey) -> bool:
    # Undefined is special because it can't be added to a DispatchKeySet, but is still considered a runtime key.
    if k == DispatchKey.Undefined:
        return True
    # None of the "backend bit" keys are real runtime keys
    if k.value <= DispatchKey.EndOfBackendKeys.value:
        return False
    # Dense/Sparse/Autograd/Quantized are not runtime keys
    if isPerBackendFunctionalityKey(k):
        return False
    # Alias keys are not runtime keys
    if isAliasDispatchKey(k):
        return False
    return True

def isAliasDispatchKey(k: DispatchKey) -> bool:
    return (k is DispatchKey.CompositeImplicitAutograd
            or k is DispatchKey.CompositeExplicitAutograd
            or k is DispatchKey.AutogradAlias)

def getAutogradKeyFromBackend(k: DispatchKey) -> DispatchKey:
    if k is DispatchKey.CPU:
        return DispatchKey.AutogradCPU
    if k is DispatchKey.XPU:
        return DispatchKey.AutogradXPU
    if k is DispatchKey.CUDA:
        return DispatchKey.AutogradCUDA
    if k is DispatchKey.XLA:
        return DispatchKey.AutogradXLA
    if k is DispatchKey.Lazy:
        return DispatchKey.AutogradLazy
    if k is DispatchKey.MLC:
        return DispatchKey.AutogradMLC
    if k is DispatchKey.HPU:
        return DispatchKey.AutogradHPU
    if k is DispatchKey.NestedTensor:
        return DispatchKey.AutogradNestedTensor
    if k is DispatchKey.PrivateUse1:
        return DispatchKey.AutogradPrivateUse1
    if k is DispatchKey.PrivateUse2:
        return DispatchKey.AutogradPrivateUse2
    if k is DispatchKey.PrivateUse3:
        return DispatchKey.AutogradPrivateUse3
    return DispatchKey.AutogradOther

def toBackendKey(k: DispatchKey) -> DispatchKey:
    # SparseCsr isn't in this list because it isn't currently considered a "per-backend" functionality
    if k in [DispatchKey.CPU, DispatchKey.AutogradCPU, DispatchKey.SparseCPU, DispatchKey.QuantizedCPU]:
        return DispatchKey.CPUBit
    if k in [DispatchKey.XPU, DispatchKey.AutogradXPU, DispatchKey.QuantizedXPU, DispatchKey.SparseXPU]:
        return DispatchKey.XPUBit
    if k in [DispatchKey.CUDA, DispatchKey.AutogradCUDA, DispatchKey.SparseCUDA, DispatchKey.QuantizedCUDA]:
        return DispatchKey.CUDABit
    if k in [DispatchKey.HIP, DispatchKey.SparseHIP]:
        return DispatchKey.HIPBit
    if k in [DispatchKey.XLA, DispatchKey.AutogradXLA]:
        return DispatchKey.XLABit
    if k in [DispatchKey.Lazy, DispatchKey.AutogradLazy]:
        return DispatchKey.LazyBit
    if k in [DispatchKey.MLC, DispatchKey.AutogradMLC]:
        return DispatchKey.MLCBit
    if k in [DispatchKey.HPU, DispatchKey.AutogradHPU]:
        return DispatchKey.HPUBit
    if k in [DispatchKey.NestedTensor, DispatchKey.AutogradNestedTensor]:
        return DispatchKey.NestedTensorBit
    if k in [DispatchKey.PrivateUse1, DispatchKey.AutogradPrivateUse1]:
        return DispatchKey.PrivateUse1Bit
    if k in [DispatchKey.PrivateUse2, DispatchKey.AutogradPrivateUse2]:
        return DispatchKey.PrivateUse2Bit
    if k in [DispatchKey.PrivateUse3, DispatchKey.AutogradPrivateUse3]:
        return DispatchKey.PrivateUse3Bit
    return k

def toFunctionalityKey(k: DispatchKey) -> DispatchKey:
    assert k.value > DispatchKey.EndOfBackendKeys.value
    assert not isPerBackendFunctionalityKey(k)
    if k in [
        DispatchKey.CPU,
        DispatchKey.CUDA,
        DispatchKey.HIP,
        DispatchKey.XLA,
        DispatchKey.Lazy,
        DispatchKey.XPU,
        DispatchKey.NestedTensor,
        DispatchKey.MLC,
        DispatchKey.HPU,
        DispatchKey.PrivateUse1,
        DispatchKey.PrivateUse2,
        DispatchKey.PrivateUse3
    ]:
        return DispatchKey.Dense
    if k in [
        DispatchKey.QuantizedCPU,
        DispatchKey.QuantizedCUDA,
        DispatchKey.QuantizedXPU,
    ]:
        return DispatchKey.Quantized
    if k in [
        DispatchKey.SparseCPU,
        DispatchKey.SparseCUDA,
        DispatchKey.SparseHIP,
        DispatchKey.SparseXPU,
    ]:
        return DispatchKey.Sparse
    if k in [
        DispatchKey.AutogradCPU,
        DispatchKey.AutogradCUDA,
        DispatchKey.AutogradXLA,
        DispatchKey.AutogradLazy,
        DispatchKey.AutogradXPU,
        DispatchKey.AutogradNestedTensor,
        DispatchKey.AutogradMLC,
        DispatchKey.AutogradHPU,
        DispatchKey.AutogradPrivateUse1,
        DispatchKey.AutogradPrivateUse2,
        DispatchKey.AutogradPrivateUse3
    ]:
        return DispatchKey.Autograd
    return k

NUM_BACKENDS = DispatchKey.EndOfBackendKeys.value + 1
